assignment5/assignment5.cpp

```diff
@@ -1,60 +1,103 @@
 #include <iostream>
 #include <vector>
+#include <map>
 
-class UnknownLanguage
-{
-	std::vector<std::string> dictionary;
+class UnknownLanguage{
+	struct Node{
+		std::string in_edges;
+		std::string out_edges;	
+	};
+	std::map<char, Node> graph;
 	std::string alphabet;
-public:
-	UnknownLanguage(std::vector<std::string> dic):dictionary(dic){}
-	std::string CalculateAlphabet(){
+	std::vector<std::string> dict;
+	void DictionaryToGraph(){
+		graph.clear();
+		for(int i=0; i<dict.size()-1; i++){
+			for(int j=0; j<dict[i].size(); j++){
 
-		alphabet.clear();
-		int before, after;
-
-		for(int i = 0; i < dictionary.size()-1; i++){
-			for(int j=0; j < dictionary[i].size(); j++){
-
-				// Comparison is done with next word, same character
-				if(j < dictionary[i+1].size() && 
-					dictionary[i].substr(0, j) == dictionary[i+1].substr(0, j)){
-
-					if(dictionary[i][j]!=dictionary[i+1][j]){
-						before = alphabet.find(dictionary[i][j]);
-						after = alphabet.find(dictionary[i+1][j]);
-
-						if(before!= std::string::npos || after!=std::string::npos){
-							if(before!=std::string::npos && after == std::string::npos){
-								// after is not in alphabet yet
-								alphabet.insert(alphabet.begin() + before+1, dictionary[i+1][j]);
-							} else if( after != std::string::npos && before == std::string::npos){
-								// before is not in alphabet yet
-								alphabet.insert(alphabet.begin() + after, dictionary[i][j]);
-							} else if(after < before){
-								// both are, but the order is incorrect
-								alphabet.erase(before, 1);
-								alphabet.insert(alphabet.begin() + after, dictionary[i][j]);
-							}
-						} else {
-							// both characters are new to the alphabet
-							alphabet.push_back(dictionary[i][j]);
-							alphabet.push_back(dictionary[i+1][j]);
-						}
+				if(j < dict[i+1].size() && 
+					dict[i].substr(0, j) == dict[i+1].substr(0, j)){
```

Comment:
Comparing the prefixes is a bit wasteful, you can compare chars one at a time until they first differ (or a string ends)